### ISSUE TITLE
GDScript: Enhancement of format specifier for Thousands Separator

### ISSUE THREAD

**Issue by just-like-that** [godotengine/godot-proposals#4971]
### Describe the project you are working on

A 2D block-base game.

### Describe the problem or limitation you are having in your project

I frequently output floats and integers for logging and debugging purposes.
The part before the decimal point might become big and therefore it's sometimes difficult to spot the magnitude of the value.
E.g. 6735465948756.3546334 [micro seconds]

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I propose an enhancement of the existing format specifier for %-syntax of "%f" and "%d" by introducing support for thousands separator ",".

By using "," in combination with "%d" and "%f" integer and float numbers get support for thousands separator.

Case-A:
Value is 4537537543.
"%d" % value produces exactly "4537537543".
By applying "%,d" % value it outputs the same BUT more readable number:
"4,537,537,543"

The same applies for floats.
6735465948756.3546334 [micro seconds]
becomes
6,735,465,948,756.35
by using "%,.2f"

[List of existing formats](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_format_string.html)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

See above.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

To support both int values as well as float wells you need more than a few lines of code.

### Is there a reason why this should be core and not an add-on in the asset library?

Format specifier are a core feature. It's just an enhancement of existing logic.

**Comment by zinnschlag** [comment]
I like the idea in principle. The one issue I see is the hardcoded , separator. Localisation is an obvious problem here. But developers may also want to use a space instead for a cleaner look and improved readability.

Note that this kind of formatting isn't used just for debugging/logging. It can also be used for preparing numbers for display in labels. In the former case we probably could get away with a hardcoded ,. in the later case not so much.

Unfortunately I don't have a solution for this issue. Obviously we can't just allow any character as a separator in a formatting string (there would be no way for the formatter to identify it as a separator).

**Comment by just-like-that** [comment]
> Note that this kind of formatting isn't used just for debugging/logging.

You are right this was just an example.
And Localisation is an topic too.

There is even the possibility of using underscore "_" as GDScript does already for the reason of displaying large numbers.

Future dependencies:

- Locale/en: 123,456,789.00
- Locale/de: 123.456.789,00

As the decimal place in local/de is actually a comma and the thousands separator is a point.

So maybe we need a more general solution which is ready for internationalization too.

**Comment by just-like-that** [comment]
> But developers may also want to use a space instead for a cleaner look and improved readability.

There is even the possibility of using underscore "_" in GDScript for the reason of displaying large numbers.

**Comment by Calinou** [comment]
- Related to
https://github.com/godotengine/godot-proposals/issues/4972 and
https://github.com/godotengine/godot-proposals/issues/2054.

I suggest adding a number formatting method, which you can call like this instead:

```gdscript
"My number: %s" % number_format(my_number)
```

This is both more reusable and more i18n-friendly. `number_format()` could take a second optional parameter specifying whether internationalization should be used (`true` by default, but can be `false` to force English localization).

**Comment by just-like-that** [comment]
> ```gdscript
> "My number: %s" % number_format(my_number)
> ```
>   
> This is both more reusable and more i18n-friendly.
In which way is this more reusable?

If want to output several numbers this way in a single string then I end up with 
"<string>" % [number_format(number-1), number_format(number-2), ... number_format(number-N)]

With support of format specifiers I simply write:
"<string>" % [number-1, number-2, ... number-N]

**Comment by just-like-that** [comment]
> You can reuse `number_format()` in situations where percent formatting is not desirable, such as the `String.format()` function (when you need named placeholders for i18n purposes).

Do I get you right on this? 

Your solution with `number_format()` looks similar to the current String.format():

`"Hi, the lottery numbers are {} {} .. {}!".number_format([number-1, number-2, ..., number-N], "{}", true)`

How do you combine both approaches in one string if you need to? 
String.format() as well as String.number_format()

**Comment by duchainer** [comment]
@just-like-that Not at my computer, but you could be escaping your {} that will host a number and have unescaped {} for the string_formatted stuff (or the other way around).

**Comment by gonza1212** [comment]
yes please, we need this feature in the core, not with custom functions

**Comment by Calinou** [comment]
@gonza1212 Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.
